pkg/providers/v1/aws.go

```diff
@@ -1503,6 +1503,23 @@ func (c *Cloud) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.No
 			}
 		}
 
+		for _, macID := range macIDs {
+			ipv6Path := path.Join("network/interfaces/macs/", macID, "ipv6s")
+			internalIPv6s, err := c.metadata.GetMetadata(ipv6Path)
+			if err != nil {
+				// Interfaces without an IPv6 address don't get the "ipv6s" data
+				klog.V(4).Infof("error querying AWS metadata for %q: %q", ipv6Path, err)
+				continue
+			}
+
+			for _, internalIPv6 := range strings.Split(internalIPv6s, "\n") {
```

Comment:
I think we should add only the 'first' address (from each ENI).

In IPv4, AWS has a strong distinction between 'primary' IPv4 and all the 'secondary' addresses (per ENI).  In IPv6, the AWS API does not make a distinction, and it's just a single list of all IPv6 addresses (per ENI).

So if/when we want to add multiple IPv6 addresses to an ENI, eg so we have VPC addresses for pods, then they will _all_ turn up in this `ipv6s` list.  I think we need to be less greedy here :) and only use the first address, so the 'rest' can be left for other purposes.  From my testing ordering is preserved, so first means naive `ipv6s[0]`.

Reply:
Sounds good to me. Done, also added a test for this case.